networks/2_module_validation.py

```diff
@@ -0,0 +1,300 @@
+        gene_id for gene_id in list_genes if gene_id in pao1_membership.index
+    ]
+)
+
+# Update length based on filtered gene list ("Genes_processed" column)
+pao1_operon["Length_processed"] = pao1_operon["Genes_processed"].str.len()
+pao1_regulon["Length_processed"] = pao1_regulon["Genes_processed"].str.len()
+
+# For each regulon/operon, select a random set of genes that are the same size at the regulon/operon
+pao1_operon["Random_Genes"] = pao1_operon["Length_processed"].apply(
+    lambda num_genes: pao1_membership.sample(num_genes).index.values
+)
+pao1_regulon["Random_Genes"] = pao1_regulon["Length_processed"].apply(
+    lambda num_genes: pao1_membership.sample(num_genes).index.values
+)
+
+pao1_operon.head()
+
+pao1_regulon.head()
+
+# ### Calculate the distribution
+
+# For each regulon/operon get the number of modules that regulon/operon genes are found in, number of modules
+# that random genes are found in
+pao1_operon["Num_operon_modules"] = pao1_operon["Genes_processed"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+pao1_operon["Num_random_modules"] = pao1_operon["Random_Genes"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+
+pao1_regulon["Num_regulon_modules"] = pao1_regulon["Genes_processed"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+pao1_regulon["Num_random_modules"] = pao1_regulon["Random_Genes"].apply(
+    lambda list_genes: pao1_membership.loc[list_genes]["module id"].nunique()
+)
+
+pao1_operon.head()
+
+pao1_regulon.head()
+
+# +
+# Format df for plotting using displot
+pao1_operon_toplot = pd.melt(
+    pao1_operon, value_vars=["Num_operon_modules", "Num_random_modules"]
+)
+pao1_regulon_toplot = pd.melt(
+    pao1_regulon, value_vars=["Num_regulon_modules", "Num_random_modules"]
+)
+
+pao1_operon_toplot.tail()
+
+
+# -
+
+def cumulative_distribution(
+    data,
+    scaled=False,
+    survival=False,
+    label="Cumulative",
+    fill=False,
+    flip=False,
+    preserve_ends=0,
+    **kwargs,
+):
+    """
+    plots cumulative (or survival) step distribution
+    adapted from https://github.com/MarvinT/morphs/blob/master/morphs/plot/utils.py
+    """
+    data = np.sort(data)
+    if survival:
+        data = data[::-1]
+    y = np.arange(data.size + 1, dtype=float)
+    if scaled:
+        y /= y[-1]
+    x = np.concatenate([data, data[[-1]]])
+    plt.step(x, y, label=label, **kwargs)
+    if fill:
+        plt.fill_between(x, y, alpha=0.5, step="pre", **kwargs)
+
+
+# +
+cumulative_distribution(
+    pao1_operon["Num_operon_modules"],
+    label="Number of modules containing operon genes",
+    color="red",
+)
+cumulative_distribution(
+    pao1_operon["Num_random_modules"],
+    label="Number of modules containing random genes",
+    color="blue",
+)
+_ = plt.legend()
+plt.title("Distribution of module counts (operon vs random genes)")
+
+scipy.stats.ks_2samp(
```

Comment:
I'm still a bit confused about exactly what this plot is showing. Is each value on the y-axis an operon? So the value at "1" on the x-axis is the number of operons where the genes in that operon are contained in a single module?

Then the increase at "2" on the x-axis would be the number of operons that are spread across 2 different modules, and so on?

If this is right, it might be helpful to label the axes with something like that (I'm not sure exactly what the labels should be, but I don't think it's super obvious what the axes are here). I guess you have some detail in the "takeaway" section at the bottom, but it may be good to describe it a little on the axes themselves too.

Reply:
I'm also still trying to wrap my head around these plots, let's see if this is more clear:

y-axis = the cumulative count of genes within operon/regulon (red) or random genes (blue).
x-axis = The number of modules that genes are contained in

Looking at the operon plot, the value at "1" on the x-axis says that there are ~3000 operon genes found in exactly 1 module, and there are ~2700 random genes found in exactly 1 module

Then the increase at "2" on the x-axis is the number of operons or random genes that are spread across 2 or 1 different modules (this is the cumulative part). There are >3000 operon genes that are found in either 1 or 2 modules. There are > 3000 random genes that are found in either 1 or 2 modules.

Then if you compare the blue and the red curves, the vertical distance between the two curves tells you how much of a shift there is between the distributions.

I will definitely add axis labels here to help. Please let me know if this explanation helps or if its still confusing.

Reply:
Ok, this helps!

Just to clarify one small point: when you say "random genes spread across 2 modules", you really mean "random gene sets (of equal size to an operon) spread across 2 modules", right? I don't think it would make sense for a single gene to be spread across multiple modules - each gene should be in exactly one module as far as I know, unless you're finding overlapping modules (but I don't think this should be the case with most hierarchical clustering implementations).

Reply:
Yes that is correct - a random set (size matched with the operon) where the genes in that set are found in 2 modules

Reply:
Sorry slight error on my part, y-axis is the cumulative number of operons/regulons/random groups NOT genes. So your initial statement was correct. Its the number of Y operons spread across X modules in the network. I'll need to update this text and labels